### ISSUE TITLE
Webpack 5 is compiling my Javascript twice

### ISSUE THREAD

**Issue by smparsons** [webpack/webpack#10789]
# Bug report

**What is the current behavior?**

I could be doing something wrong but it appears that Webpack 5 is behaving differently than Webpack 4 when compiling. 

Specifically in Webpack 5, when I compile and run with express and webpack-dev-middleware, I see in the console that Webpack is initially compiling my Javascript twice, however I don't see this issue in Webpack 4.

This is my console output:

```
<i> [webpack-dev-middleware] Hash: bd9dbee05d4205acc7b6
<i> Version: webpack 5.0.0-beta.15
<i> Time: 118 ms
<i> Built at: 2020-04-27 12:010:31
<i>                                    Asset       Size
<i>                               index.html  211 bytes  [emitted]
<i>     scripts/main.bd9dbee05d4205acc7b6.js  340 bytes  [emitted] [immutable]  [name: main]
<i> scripts/main.bd9dbee05d4205acc7b6.js.map  229 bytes  [emitted] [dev]        [name: (main)]
<i> Entrypoint main = scripts/main.bd9dbee05d4205acc7b6.js (scripts/main.bd9dbee05d4205acc7b6.js.map)
<i> ./index.js 31 bytes [built]
<i> [webpack-dev-middleware] Compiled successfully.
<i> [webpack-dev-middleware] Compiling...
<i> [webpack-dev-middleware] Hash: 6aa51fc752fd10b023b2
<i> Version: webpack 5.0.0-beta.15
<i> Time: 11 ms
<i> Built at: 2020-04-27 12:010:31
<i>                                    Asset       Size
<i>                               index.html  211 bytes  [emitted]
<i>     scripts/main.6aa51fc752fd10b023b2.js  340 bytes  [emitted] [immutable]  [name: main]
<i> scripts/main.6aa51fc752fd10b023b2.js.map  256 bytes  [emitted] [dev]        [name: (main)]
<i> Entrypoint main = scripts/main.6aa51fc752fd10b023b2.js (scripts/main.6aa51fc752fd10b023b2.js.map)
<i> ./index.js 31 bytes
<i> [webpack-dev-middleware] Compiled successfully.
```

**If the current behavior is a bug, please provide the steps to reproduce.**

Please see my minimal repo that reproduces the issue: https://github.com/smparsons/webpack-builds-twice-issue

The steps to reproduce are in the README file.

**What is the expected behavior?**

I would expect that Webpack 5 would compile my Javascript only once. That's at least how Webpack 4 behaves.

**Other relevant information:**
webpack version:  5.0.0-beta.15
Node.js version: 11.10.1
Operating System: macOS Catalina Version 10.15
Additional tools:

**Comment by sokra** [comment]
@jantimon Removing [this line](https://github.com/jantimon/html-webpack-plugin/blob/96e0ae607afa7295c40d0a5073c791189def6337/lib/webpack5/file-watcher-api.js#L61) fixes the problem.

`.../webpack-builds-twice-issue/node_modules/html-webpack-plugin/lib/package.json` is a `missingDependency` but added to `fileDependencies` too. The watcher do not find the file and as it had no watcher attached initially, it cannot know when the file were removed. So to be safe it has to trigger a change event, which triggers a compilation.

In the second run it has a watcher attached and is able to tell that there were no change to this file.

**Comment by jantimon** [comment]
Released as html-webpack-plugin@4.2.1

**Comment by evilebottnawi** [comment]
@jantimon Thanks!

Let's close the issue

@smparsons feel free to leave feedback or open a new issue if you still have problems

**Comment by janlent1** [comment]
@evilebottnawi now I am having this issue again with "webpack": "5.1.3"... I am upgrading from webpack 4 and unfortunately, I cannot post my config, but when I hook into the compiler hook invalid, i only get the timestamp and an undefined as the second argument which should be the file that is changed or not valid anymore afaik.
It also talks about orphan modules after the first build and then the second build has none and some different processed modules. The main build takes a longer while(8-10 mins), the second only a minute max. I know it takes long, but that is because it is quite a big repo and many custom deps, custom plugins, loaders, ....
I know this is difficult to deal with, but maybe I could give you more webpack specific output that is not necessarily revealing the configuration.
Thanks and Best,
Jan
